pkg/task/taskset.go

```diff
@@ -209,46 +206,81 @@ func (s *TaskSet) sortTaskGraph() *SortTaskGraphResult {
 			// Gathers the cause of the failure.
 			missingTaskIdsInMap := map[string]interface{}{}
 			for taskId := range nonResolvedTasksMap {
-				for source := range currentMissingTaskDependencies[taskId] {
-					missingTaskIdsInMap[source] = struct{}{}
+				for dependency := range currentMissingTaskDependencies[taskId] {
+					missingTaskIdsInMap[dependency] = struct{}{}
 				}
 			}
 			for _, task := range nonResolvedTasksMap {
 				delete(missingTaskIdsInMap, task.UntypedID().ReferenceIDString())
 			}
 
-			// When there were no task runnable only with the missing sources,
-			// the task graph shape is at least cyclic.
-			hasCyclicDependencies := true
-			for taskId := range nonResolvedTasksMap {
-				canBeNextStartingPoint := true
-				for source := range currentMissingTaskDependencies[taskId] {
-					if _, exist := missingTaskIdsInMap[source]; !exist {
-						canBeNextStartingPoint = false
-						break
-					}
-				}
-				if canBeNextStartingPoint {
-					hasCyclicDependencies = false
-					break
-				}
-			}
-
-			if !hasCyclicDependencies {
-				for _, task := range nonResolvedTasksMap {
-					delete(missingTaskIdsInMap, task.UntypedID().ReferenceIDString())
-				}
-			}
-
 			missingSources := []taskid.UntypedTaskReference{}
 			for source := range missingTaskIdsInMap {
 				missingSources = append(missingSources, taskid.NewTaskReference[any](source))
 			}
 
+			if len(missingSources) == 0 {
```

Comment:
I guess this part is actually detecting the cyclic and building the path, however, this is complicated enough.

Can we define this part as another function so that we can focus 'sort' part?

Reply:
Done.